pkg/loop/client.go

```diff
@@ -0,0 +1,154 @@
+}
+
+func (a *atomicClient) store(cc *grpc.ClientConn) { a.cc.Store(cc) }
+
+func (a *atomicClient) Invoke(ctx context.Context, method string, args interface{}, reply interface{}, opts ...grpc.CallOption) error {
+	return a.cc.Load().Invoke(ctx, method, args, reply, opts...)
+}
+
+func (a *atomicClient) NewStream(ctx context.Context, desc *grpc.StreamDesc, method string, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+	return a.cc.Load().NewStream(ctx, desc, method, opts...)
+}
+
+var _ grpc.ClientConnInterface = (*clientConn)(nil)
+
+// newClientFn returns a new client connection id to dial, and a set of resource dependencies to close.
+type newClientFn func(context.Context) (id uint32, deps resources, err error)
+
+// clientConn is a [grpc.ClientConnInterface] backed by a [*grpc.ClientConn] which can be recreated and swapped out
+// via the provided [newClientFn].
+type clientConn struct {
+	*lggrBroker
+	newClient newClientFn
+	name      string
+
+	mu   sync.RWMutex
+	deps resources
+	cc   *grpc.ClientConn
+}
+
+func (lb *lggrBroker) newClientConn(name string, newClient newClientFn) *clientConn {
+	return &clientConn{
+		lggrBroker: lb.named(name),
+		newClient:  newClient,
+		name:       name,
+	}
+}
+
+func (c *clientConn) Invoke(ctx context.Context, method string, args interface{}, reply interface{}, opts ...grpc.CallOption) error {
+	c.mu.RLock()
+	cc := c.cc
+	c.mu.RUnlock()
+
+	if cc == nil {
+		cc = c.refresh(ctx, nil)
+	}
+	for cc != nil {
+		err := cc.Invoke(ctx, method, args, reply, opts...)
+		if err == nil || !isErrTerminal(err) {
+			return err
+		}
+		cc = c.refresh(ctx, cc)
+	}
+	return context.Cause(ctx)
+}
+
+func (c *clientConn) NewStream(ctx context.Context, desc *grpc.StreamDesc, method string, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+	c.mu.RLock()
+	cc := c.cc
+	c.mu.RUnlock()
+
+	if cc == nil {
+		cc = c.refresh(ctx, nil)
+	}
+	for cc != nil {
+		s, err := cc.NewStream(ctx, desc, method, opts...)
+		if err == nil || !isErrTerminal(err) {
+			return s, err
+		}
+		cc = c.refresh(ctx, cc)
+	}
+	return nil, context.Cause(ctx)
+}
+
+func (c *clientConn) refresh(ctx context.Context, orig *grpc.ClientConn) *grpc.ClientConn {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.cc != orig {
+		return c.cc
+	}
+	if c.cc != nil {
+		if err := c.cc.Close(); err != nil {
+			c.lggr.Errorw("Client close failed", "err", err)
+		}
+		c.closeAll(c.deps...)
+	}
+
+	try := func() bool {
+		c.lggr.Debug("Client refresh")
+		id, deps, err := c.newClient(ctx)
+		if err != nil {
+			c.lggr.Errorw("Client refresh failed", "err", err)
+			c.closeAll(deps...)
+			return false
+		}
+		c.deps = deps
+
+		lggr := logger.With(c.lggr, "id", id)
+		lggr.Debug("Client dial")
+		c.cc, err = c.broker.Dial(id)
+		if err != nil {
+			err = ErrConnDial{Name: c.name, ID: id, Err: err}
+			lggr.Errorw("Client dial failed", "err", err)
+			c.closeAll(c.deps...)
+			return false
+		}
+		return true
+	}
+
+	const maxWait = 5 * time.Second
```

Comment:
why? where did this come from

Reply:
i don't follow spam vs silence. i thought maxWait was a timeout. re-reading, does this loop forever if try() always returns false?

Reply:
Yeah, loops until connection is established or context expires.

Reply:
i see. it warrants a comment -- i wouldn't expect NewStream to block forever under some circumstance. that is, it's very dangerous for a caller to pass a context with a timeout or deadline

Reply:
Indeed, but we have no choice currently since we cannot even return errors fully in some cases. If we can get `error`s introduced to a few more libocr methods, then we'll be able to consider returning early from here, but that won't necessarily 'just work' - we have to make sure the code up the stack is prepared to retry on it's own.

Reply:
libocr defines many of the interfaces that make up this API.